File: 63 (Unique Paths II)/python/Solution.py
class Solution:
    def uniquePathsWithObstacles(self, grid: List[List[int]]) -> int:

        n,m = len(grid), len(grid[0])
        if (grid[0][0]==1 or grid[n-1][m-1]==1): return 0


        @cache
        def solve(i,j):
            if i==n-1 and j==m-1: return 1

            if i>=n or j >= m or grid[i][j]==1: return 0

            return solve(i+1,j) + solve(i,j+1)
    
        return solve(0,0)

File: 63 (Unique Paths II)/java/Solution.java
class Solution {
    public int uniquePathsWithObstacles(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        if (matrix[0][0]==1 || matrix[n-1][m-1]==1) return 0;

        int[][] dp = new int[n][m];

        dp[0][0] = 1;
        for(int i = 0;i<n;i++){
            for(int j =0;j<m;j++){
                if (matrix[i][j]==1) continue;

                if (i-1>=0) dp[i][j] += dp[i-1][j];
                if (j-1>=0) dp[i][j] += dp[i][j-1];

            }
        }
        return dp[n-1][m-1];
    }
}